common/sysinfo_linux.go

```diff
@@ -0,0 +1,154 @@
+//go:build linux || darwin
+// +build linux darwin
+
+package common
+
+import (
+	"bufio"
+	"fmt"
+	"os/exec"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"golang.org/x/sys/unix"
+)
+
+// GetMemInfo returns the MemAvailable on linux system.
+// MemAvailable field added after 3.14+ kernel only. So for kernel version before that we need to give rough estimate.
+//
+// TODO: Estimating MemAvailable on kernel before 3.14.
+func GetMemInfo() (int64, error) {
+
+	// command to get the Available Memory
+	cmdStr := `cat /proc/meminfo`
+	cmd := exec.Command("sh", "-c", cmdStr)
+
+	stdOut, err := cmd.StdoutPipe()
+	if err != nil {
+		err := fmt.Errorf("GetMemInfo failed with error: %v", err)
+		return 0, err
+	}
+
+	// Flow will be like this, set the scanner to stdOut and start the cmd.
+	scanner := bufio.NewScanner(stdOut)
+	err = cmd.Start()
+	if err != nil {
+		err := fmt.Errorf("GetMemInfo failed with error: %v", err)
+		return 0, err
+	}
+
+	// Set the split function for the scanning operation.
+	scanner.Split(bufio.ScanLines)
+
+	// Scan the stdOutput.
+	for scanner.Scan() {
+		if strings.HasPrefix(scanner.Text(), "MemAvailable") {
+			var multipler int64
+			var result int
+			re := regexp.MustCompile(` +`)
+			value := strings.Split(string(re.ReplaceAll([]byte(scanner.Text()), []byte(" "))), " ")[1]
+			multiplerStr := strings.Split(string(re.ReplaceAll([]byte(scanner.Text()), []byte(" "))), " ")[2]
+
+			if multiplerStr == "kB" {
+				multipler = 1024
+			} else {
+				// "/proc/meminfo" output always in kB only. If we are getting different string, something wrong.
+				err := fmt.Errorf("Meminfo value is not in kB")
+				return 0, err
+			}
+
+			if result, err = strconv.Atoi(value); err != nil {
+				return 0, err
+			}
+			return int64(result) * int64(multipler), nil
+		}
+	}
+
+	if err = cmd.Wait(); err != nil {
+		err := fmt.Errorf("GetMemInfo failed with error: %v", err)
+		return 0, err
+	}
+
```

Comment:
this should be a failure return since we couldn't get valid available memory

Reply:
Yes, but we need add the memfree code here .